internal/grpcwrap/provider6.go

```diff
@@ -891,6 +918,219 @@ func (p *provider6) ConfigureStateStore(ctx context.Context, req *tfplugin6.Conf
 	panic("not implemented")
 }
 
+func (p *provider6) PlanAction(_ context.Context, req *tfplugin6.PlanAction_Request) (*tfplugin6.PlanAction_Response, error) {
+	resp := &tfplugin6.PlanAction_Response{}
+
+	actionSchema, ok := p.schema.Actions[req.ActionType]
+	if !ok {
+		return nil, fmt.Errorf("action schema not found for action %q", req.ActionType)
+	}
+
+	ty := actionSchema.ConfigSchema.ImpliedType()
+	configVal, err := decodeDynamicValue6(req.Config, ty)
+	if err != nil {
+		resp.Diagnostics = convert.AppendProtoDiag(resp.Diagnostics, err)
+		return resp, nil
+	}
+
+	linkedResouceSchemas := actionSchema.LinkedResources()
+	inputLinkedResources := make([]providers.LinkedResourcePlanData, 0, len(req.LinkedResources))
+
+	for i, lr := range req.LinkedResources {
+		resourceSchema, ok := p.schema.ResourceTypes[linkedResouceSchemas[i].TypeName]
+		if !ok {
+			return nil, fmt.Errorf("linked resource schema not found for type %q in action %q", linkedResouceSchemas[i].TypeName, req.ActionType)
+		}
+
+		linkedResourceTy := resourceSchema.Body.ImpliedType()
+		linkedResourceIdentityTy := resourceSchema.Identity.ImpliedType()
+
+		priorState, err := decodeDynamicValue6(lr.PriorState, linkedResourceTy)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode prior state for linked resource #%d (%q) in action %q: %w", i+1, linkedResouceSchemas[i].TypeName, req.ActionType, err)
```

Comment:
Is `i+1` the correct thing to show the user here? I thought terraform generally always used [0] for the first item in a list and it seems like it'd be odd if we switched that just for actions (I could also be wrong about that given this context!)